File: solutions/php/abc198_e.php
<?php

declare(strict_types=1);

$integers = fn () => array_map('intval', explode(' ', fgets(STDIN)));
$printn = fn ($value) => print($value . "\n");

class Graph
{
  private $N;
  private $M;
  private $list;

  public function __construct(int $N, int $M)
  {
    $this->N = $N;
    $this->M = $M;
    $this->list = $this->inputAdjacentList($N, $M);
  }

  private $answers = [];
  private $counts;

  public function solve(array $colors)
  {
    $this->counts = array_fill(1, (int)1e5, 0);
    $this->dfs(0, -1, $colors);
    sort($this->answers);

    return $this->answers;
  }

  public function dfs(int $v, int $parent, array $colors)
  {
    $this->counts[$colors[$v]]++;
    if ($this->counts[$colors[$v]] === 1) $this->answers[] = $v + 1;

    foreach ($this->list[$v] as $u) if ($u != $parent) {
      $this->dfs($u, $v, $colors);
    }

    $this->counts[$colors[$v]]--;
  }

  private function inputAdjacentList(int $N, int $M)
  {
    $list = array_fill(0, $N, []);
    for ($i = 0; $i < $M; $i++) {
      [$a, $b] = fscanf(STDIN, "%d%d");
      $list[$a - 1][] = $b - 1;
      $list[$b - 1][] = $a - 1;
    }
    return $list;
  }
}

[$N] = $integers();
$colors = $integers();
$graph = new Graph($N, $N - 1);

$answers = $graph->solve($colors);
$printn(implode("\n", $answers));

File: solutions/ruby/abc198_e.rb
class Graph
  def initialize(n, m)
    @n, @m = n, m
    @list = inputAdjacentList(n, m)
  end

  def solve(colors)
    @answers = []
    @counts = Array.new(10 ** 5 + 1, 0)
    dfs(0, -1, colors)

    @answers
  end

  def dfs(v, parent, colors)
    @counts[colors[v]] += 1
    @answers << v + 1 if @counts[colors[v]] == 1

    @list[v].each do |u|
      dfs(u, v, colors) if u != parent
    end

    @counts[colors[v]] -= 1
  end

  private

  def inputAdjacentList(n, m)
    list = Array.new(n) { [] }
    m.times do
      a, b = gets.split(" ").map(&:to_i)
      list[a - 1] << b - 1
      list[b - 1] << a - 1
    end
    list
  end
end

n = gets.to_i
colors = gets.split(" ").map(&:to_i)
graph = Graph.new(n, n - 1)

answers = graph.solve(colors)
puts answers.sort.join("\n")
